contracts/AckToEarn.sol

```diff
@@ -0,0 +1,124 @@
+    */
+    function sendBid(string memory message, address recipient) external payable {
+        require(recipient != address(0), "Recipient cannot be the zero address");
+
+        // The recipient gets 90% of the bid amount
+        uint recipientAmount = (msg.value * 90) / 100;
+        uint newBidId = bidIds.current();
+        Bid memory bid = Bid({
+            id: newBidId,
+            bidder: msg.sender,
+            recipient: recipient,
+            amount: msg.value,
+            message: message,
+            timestamp: block.timestamp,
+            claimed: false,
+            exists: true
+        });
+
+        bids.push(bid);
+
+        bidderBids[msg.sender][newBidId] = bid;
+        recipientBids[recipient][newBidId] = bid;
+
+        emit NewBid(msg.sender, recipient, msg.value);
+    }
+
+    /**
+    * @notice Returns all bids that have ever been made on the platform
+    */
+    function getBids() external view returns(Bid[] memory) {
+        return bids;
+    }
+
+    /**
+    * @notice Reclaims a specific set of expired bids that the bidder has made.
+    */
+    function reclaimBids(uint[] memory bidIdsToReclaim) external nonReentrant {
+        uint totalReclaimAmount = 0;
+
+        for (uint i = 0; i < bidIdsToReclaim.length; i++)
+        {
+            uint bidId = bidIdsToReclaim[i];
+            Bid memory bid = bidderBids[msg.sender][bidId];
+
+            if (!bid.exists) {
+                continue;
+            }
+            // If the recipient has already claimed the bid, don't allow the balance to be reclaimed
+            if (bid.claimed) {
+                continue;
+            }
+            // If the bid has not expired yet then don't allow the balance to be reclaimed
+            if (!isBidExpired(bidId, msg.sender)) {
+                continue;
+            }
+
+            uint bidReclaimAmount = (bid.amount * 90) / 100;
+            totalReclaimAmount += bidReclaimAmount;
+            bidderBids[msg.sender][bidId].claimed = true;
+        }
+
+        require(totalReclaimAmount > 0, "No ether to reclaim.");
+
+        (bool success, ) = (msg.sender).call{value: totalReclaimAmount}("");
+
+        require(success, "Failed to send ether.");
+
+        emit ReclaimBid(msg.sender, totalReclaimAmount);
+    }
+
+    /**
+    * @notice Check if bid is expired
+    * @dev Returns true if the bid cannot be found
+    */
+    function isBidExpired(uint bidId, address account) public view returns(bool) {
+        bool isBidderAccount = bidderBids[account][bidId].exists == true;
+        bool isRecipientAccount = recipientBids[account][bidId].exists == true;
+
+        if (isBidderAccount) {
+            return (block.timestamp >= (bidderBids[account][bidId].timestamp + bidExpiryThreshold));
+        }
+        if (isRecipientAccount) {
+            return (block.timestamp >= (recipientBids[account][bidId].timestamp + bidExpiryThreshold));
+        }
+
+        return true;
```

Comment:
we should never get here in theory, right? maybe we should throw an error?

Reply:
We would get here if the bid cannot be found. Since we are allowing multiple bids to be reclaimed at once, we are calling this function multiple times in the transaction to check if each bid exists. If we threw an error here then the entire transaction would be reverted if 1 bid doesn't exist. I didn't think that made for a great user experience but I certainly could throw an exception, if you insist...

Reply:
Minimally, I think we should change this to an `internal` function.

Reply:
Ah gotcha, I'm fine with not throwing an exception then